{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip, os, sys\n",
    "import argparse\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from _include import log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ass_f_dir='features/'\n",
    "ass_s_dir='sequences/'\n",
    "delim=';'\n",
    "match_str='16S ribosomal RNA.*'\n",
    "len_min=500\n",
    "len_max=2000 \n",
    "overwrite=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_processed_assembly_ids(\n",
    "\t\tassembly_df,\n",
    "\t\tfasta_filename,\n",
    "\t\tfasta_delim=';',\n",
    "\t\toverwrite=False\n",
    "\t):\n",
    "\tprocessed_assembly_ids = set([])\n",
    "\tlast_assid = ''\n",
    "\tstart_index = 0\n",
    "\tif os.path.isfile(fasta_filename) and not overwrite:\n",
    "\t\tlog('* output '+fasta_filename+' already exists: ', end='')\n",
    "\t\tlog('appending to file...', end='')\n",
    "\t\twith open(fasta_filename, 'r') as fasta_out:\n",
    "\t\t\tn_processed_ids = 0\n",
    "\t\t\tfor record in SeqIO.parse(fasta_out, 'fasta'):\n",
    "\t\t\t\tlast_assid = record.id.strip().split(fasta_delim)[0]\n",
    "\t\t\t\tprocessed_assembly_ids |= set([last_assid])\n",
    "\t\t\t\tn_processed_ids += 1\n",
    "\t\ttry:\n",
    "\t\t\tlast_checked_index = assembly_df['#assembly_accession'] == last_assid\n",
    "\t\t\tstart_index = assembly_df.index[last_checked_index].tolist()[0]                    \n",
    "\t\texcept:\n",
    "\t\t\tstart_index = 0\n",
    "\t\tlog('OK. Added % 5d sequences.' % (n_processed_ids))\n",
    "\treturn (processed_assembly_ids, start_index)\n",
    "\n",
    "def load_gff_feature_table(filename):\n",
    "\tcolumns = {\n",
    "\t\t'assembly': str,\n",
    "\t\t'genomic_accession':str,\n",
    "\t\t'start': int,\n",
    "\t\t'end': int, \n",
    "\t\t'strand': str,\n",
    "\t\t'name': str,\n",
    "\t\t'attributes': str\n",
    "\t}\n",
    "\treturn (\n",
    "\t\tpd.read_csv(\n",
    "\t\t\tfilename,\n",
    "\t\t\tsep='\\t',\n",
    "\t\t\tcompression='gzip',\n",
    "\t\t\tusecols=list(columns.keys()),\n",
    "\t\t\tdtype=columns,\n",
    "\t\t\tkeep_default_na=False,\n",
    "\t\t)\t\t\t\t\t\n",
    "\t)\n",
    "\n",
    "def build_file_path(ftp_path, base_path, type):\n",
    "\tsuffix = {'features': 'feature_table.txt', 'sequences': 'genomic.fna'}\n",
    "\treturn os.path.join(\n",
    "\t\tbase_path,\n",
    " \t\tos.path.basename(ftp_path) + '_' + suffix[type] + '.gz',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ass_sub = '../data/archaea_assembly_summary_filter_2023-07-13.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>ftp_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009428885.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_009729015.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_003201835.2</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_009729545.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1455</th>\n",
       "      <td>GCF_001316265.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1456</th>\n",
       "      <td>GCF_001316285.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1457</th>\n",
       "      <td>GCF_001748385.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1458</th>\n",
       "      <td>GCF_902786095.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>GCF_948444175.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1460 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #assembly_accession                                           ftp_path\n",
       "0        GCF_009428885.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1        GCF_009729015.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "2        GCF_003201835.2  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "3        GCF_000213215.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "4        GCF_009729545.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "...                  ...                                                ...\n",
       "1455     GCF_001316265.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1456     GCF_001316285.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1457     GCF_001748385.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1458     GCF_902786095.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...\n",
       "1459     GCF_948444175.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...\n",
       "\n",
       "[1460 rows x 2 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fa_out_name = '../test.fa'\n",
    "\n",
    "assembly_df = pd.read_csv(\n",
    "\tass_sub,\n",
    "\tsep='\\t',\n",
    "\tusecols=['#assembly_accession', 'ftp_path'],\n",
    "\tdtype={'#assembly_accession': str, 'ftp_path': str}\n",
    ")\n",
    "\n",
    "processed_assembly_ids, start_index = add_processed_assembly_ids(assembly_df, fa_out_name)\n",
    "assembly_df = assembly_df[~assembly_df['#assembly_accession'].isin(processed_assembly_ids)]\n",
    "assembly_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#assembly_accession                                      GCF_000213215.1\n",
       "ftp_path               https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "features_path          ../data/features/GCF_000213215.1_ASM21321v1_fe...\n",
       "sequences_path         ../data/sequences/GCF_000213215.1_ASM21321v1_g...\n",
       "features_exist                                                      True\n",
       "sequences_exist                                                     True\n",
       "Name: 3, dtype: object"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_df.loc[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_paths(ftp_path, base_paths):\n",
    "\tpath_prefix = os.path.basename(ftp_path)\n",
    "\tfeatures_paths = os.path.join(base_paths['features'], path_prefix + '_feature_table.txt.gz')\n",
    "\tsequences_paths = os.path.join(base_paths['sequences'], path_prefix + '_genomic.fna.gz')\n",
    "\treturn pd.Series(\n",
    "\t\t[features_paths, sequences_paths],\n",
    "\t\tindex=['features_path', 'sequences_path']\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "assembly_df[['features_path', 'sequences_path']] = assembly_df['ftp_path'].apply(\n",
    "    build_paths,\n",
    "    base_paths = {\n",
    "        'features': '../data/'+ass_f_dir,\n",
    "        'sequences': '../data/'+ass_s_dir,\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def file_exists(filename):\n",
    "\treturn os.path.exists(filename) and os.path.getsize(filename) > 0\n",
    "\n",
    "def filter_missing_files(assembly_df):\n",
    "\tassembly_df['features_exist'] = assembly_df['features_path'].apply(file_exists)\n",
    "\tassembly_df['sequences_exist'] = assembly_df['sequences_path'].apply(file_exists)\n",
    "\treturn (\n",
    "\t\tassembly_df\n",
    "\t\t.query('features_exist == True & sequences_exist == True')\n",
    "\t\t.drop(columns=['features_exist', 'sequences_exist'])\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>features_path</th>\n",
       "      <th>sequences_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009428885.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009428885.1_ASM942888v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009428885.1_ASM942888v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_009729015.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009729015.1_ASM972901v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009729015.1_ASM972901v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_003201835.2</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_003201835.2_ASM320183v2_f...</td>\n",
       "      <td>../data/sequences/GCF_003201835.2_ASM320183v2_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000213215.1_ASM21321v1_fe...</td>\n",
       "      <td>../data/sequences/GCF_000213215.1_ASM21321v1_g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_009729545.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009729545.1_ASM972954v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009729545.1_ASM972954v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1445</th>\n",
       "      <td>GCF_014962245.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_014962245.1_ASM1496224v1_...</td>\n",
       "      <td>../data/sequences/GCF_014962245.1_ASM1496224v1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446</th>\n",
       "      <td>GCF_000092185.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000092185.1_ASM9218v1_fea...</td>\n",
       "      <td>../data/sequences/GCF_000092185.1_ASM9218v1_ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1447</th>\n",
       "      <td>GCF_000148385.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000148385.1_ASM14838v1_fe...</td>\n",
       "      <td>../data/sequences/GCF_000148385.1_ASM14838v1_g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1449</th>\n",
       "      <td>GCF_001316005.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_001316005.1_ASM131600v1_f...</td>\n",
       "      <td>../data/sequences/GCF_001316005.1_ASM131600v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1450</th>\n",
       "      <td>GCF_000190315.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000190315.1_ASM19031v1_fe...</td>\n",
       "      <td>../data/sequences/GCF_000190315.1_ASM19031v1_g...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1450 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #assembly_accession                                           ftp_path  \\\n",
       "0        GCF_009428885.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1        GCF_009729015.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "2        GCF_003201835.2  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3        GCF_000213215.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "4        GCF_009729545.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "...                  ...                                                ...   \n",
       "1445     GCF_014962245.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1446     GCF_000092185.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1447     GCF_000148385.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1449     GCF_001316005.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1450     GCF_000190315.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "\n",
       "                                          features_path  \\\n",
       "0     ../data/features/GCF_009428885.1_ASM942888v1_f...   \n",
       "1     ../data/features/GCF_009729015.1_ASM972901v1_f...   \n",
       "2     ../data/features/GCF_003201835.2_ASM320183v2_f...   \n",
       "3     ../data/features/GCF_000213215.1_ASM21321v1_fe...   \n",
       "4     ../data/features/GCF_009729545.1_ASM972954v1_f...   \n",
       "...                                                 ...   \n",
       "1445  ../data/features/GCF_014962245.1_ASM1496224v1_...   \n",
       "1446  ../data/features/GCF_000092185.1_ASM9218v1_fea...   \n",
       "1447  ../data/features/GCF_000148385.1_ASM14838v1_fe...   \n",
       "1449  ../data/features/GCF_001316005.1_ASM131600v1_f...   \n",
       "1450  ../data/features/GCF_000190315.1_ASM19031v1_fe...   \n",
       "\n",
       "                                         sequences_path  \n",
       "0     ../data/sequences/GCF_009428885.1_ASM942888v1_...  \n",
       "1     ../data/sequences/GCF_009729015.1_ASM972901v1_...  \n",
       "2     ../data/sequences/GCF_003201835.2_ASM320183v2_...  \n",
       "3     ../data/sequences/GCF_000213215.1_ASM21321v1_g...  \n",
       "4     ../data/sequences/GCF_009729545.1_ASM972954v1_...  \n",
       "...                                                 ...  \n",
       "1445  ../data/sequences/GCF_014962245.1_ASM1496224v1...  \n",
       "1446  ../data/sequences/GCF_000092185.1_ASM9218v1_ge...  \n",
       "1447  ../data/sequences/GCF_000148385.1_ASM14838v1_g...  \n",
       "1449  ../data/sequences/GCF_001316005.1_ASM131600v1_...  \n",
       "1450  ../data/sequences/GCF_000190315.1_ASM19031v1_g...  \n",
       "\n",
       "[1450 rows x 4 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_missing_files(assembly_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>761380</td>\n",
       "      <td>762036</td>\n",
       "      <td>+</td>\n",
       "      <td>16S ribosomal RNA methyltransferase A</td>\n",
       "      <td>657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3035</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>1296061</td>\n",
       "      <td>1297560</td>\n",
       "      <td>+</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>1500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             assembly genomic_accession    start      end strand  \\\n",
       "1739  GCF_000213215.1       NC_015518.1   761380   762036      +   \n",
       "3035  GCF_000213215.1       NC_015518.1  1296061  1297560      +   \n",
       "\n",
       "                                       name  length  \n",
       "1739  16S ribosomal RNA methyltransferase A     657  \n",
       "3035                      16S ribosomal RNA    1500  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find all genomic_accession for all 16S rRNA\n",
    "\n",
    "# Load GFF, find 16S records, filter by length\n",
    "features_filename = assembly_df.loc[3, 'features_path']\n",
    "sequences_filename = assembly_df.loc[3, 'sequences_path']\n",
    "\n",
    "def load_16s_records(features_filename, length_thresholds={'min': 500, 'max': 2000}):\n",
    "\t# match_str = '16S ribosomal RNA(?!.*methyl).*'\n",
    "\tmatch_str = '16S ribosomal RNA.*'\n",
    "\tdf = load_gff_feature_table(features_filename)\n",
    "\tfeatures_df = df[df['name'].str.contains(match_str)].copy()\n",
    "\tfeatures_df['length'] = features_df['end'] - features_df['start'] + 1\n",
    "\tfeatures_df = features_df.query(\n",
    "\t\tf'length >= {length_thresholds[\"min\"]} & length <= {length_thresholds[\"max\"]}'\n",
    "\t)\n",
    "\treturn features_df.drop(columns=['attributes'])\n",
    "\n",
    "features_16s_df = load_16s_records(features_filename)\n",
    "features_16s_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_fasta_metadata(df, delim=';'):\n",
    "\treturn (\n",
    "\t\t'>'\n",
    "\t\t+ str(df['assembly'])\n",
    "\t\t+ delim + str(df['genomic_accession'])\n",
    "\t\t+ delim + 'loc:' + str(df['start']) + ',' + str(df['end']-1)\n",
    "\t\t+ delim + 'strand:' + df['strand']\n",
    "\t\t+ delim + 'length' + str(df['length'])\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>metadata</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>761380</td>\n",
       "      <td>762036</td>\n",
       "      <td>-</td>\n",
       "      <td>16S ribosomal RNA methyltransferase A</td>\n",
       "      <td>657</td>\n",
       "      <td>&gt;GCF_000213215.1;NC_015518.1;loc:761380,762035...</td>\n",
       "      <td>TTATACATAATTTAATCCCATTAAATTCAATAGTTCTATAACTTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3035</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>1296061</td>\n",
       "      <td>1297560</td>\n",
       "      <td>+</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>1500</td>\n",
       "      <td>&gt;GCF_000213215.1;NC_015518.1;loc:1296061,12975...</td>\n",
       "      <td>TTCCGGTTGATCCTGCCGGACCCGACCGCTATGGGGGTAGGGCTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             assembly genomic_accession    start      end strand  \\\n",
       "1739  GCF_000213215.1       NC_015518.1   761380   762036      -   \n",
       "3035  GCF_000213215.1       NC_015518.1  1296061  1297560      +   \n",
       "\n",
       "                                       name  length  \\\n",
       "1739  16S ribosomal RNA methyltransferase A     657   \n",
       "3035                      16S ribosomal RNA    1500   \n",
       "\n",
       "                                               metadata  \\\n",
       "1739  >GCF_000213215.1;NC_015518.1;loc:761380,762035...   \n",
       "3035  >GCF_000213215.1;NC_015518.1;loc:1296061,12975...   \n",
       "\n",
       "                                               sequence  \n",
       "1739  TTATACATAATTTAATCCCATTAAATTCAATAGTTCTATAACTTGA...  \n",
       "3035  TTCCGGTTGATCCTGCCGGACCCGACCGCTATGGGGGTAGGGCTAA...  "
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_16s_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>metadata</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>761380</td>\n",
       "      <td>762036</td>\n",
       "      <td>-</td>\n",
       "      <td>16S ribosomal RNA methyltransferase A</td>\n",
       "      <td>657</td>\n",
       "      <td>&gt;GCF_000213215.1;NC_015518.1;loc:761380,762035...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3035</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>1296061</td>\n",
       "      <td>1297560</td>\n",
       "      <td>+</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>1500</td>\n",
       "      <td>&gt;GCF_000213215.1;NC_015518.1;loc:1296061,12975...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             assembly genomic_accession    start      end strand  \\\n",
       "1739  GCF_000213215.1       NC_015518.1   761380   762036      -   \n",
       "3035  GCF_000213215.1       NC_015518.1  1296061  1297560      +   \n",
       "\n",
       "                                       name  length  \\\n",
       "1739  16S ribosomal RNA methyltransferase A     657   \n",
       "3035                      16S ribosomal RNA    1500   \n",
       "\n",
       "                                               metadata sequence  \n",
       "1739  >GCF_000213215.1;NC_015518.1;loc:761380,762035...           \n",
       "3035  >GCF_000213215.1;NC_015518.1;loc:1296061,12975...           "
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_16s_df['metadata'] = features_16s_df.apply(create_fasta_metadata, axis=1)\n",
    "features_16s_df['sequence'] = ''\n",
    "features_16s_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_paths(ftp_path, base_paths):\n",
    "\tpath_prefix = os.path.basename(ftp_path)\n",
    "\tfeatures_paths = os.path.join(base_paths['features'], path_prefix + '_feature_table.txt.gz')\n",
    "\tsequences_paths = os.path.join(base_paths['sequences'], path_prefix + '_genomic.fna.gz')\n",
    "\treturn pd.Series(\n",
    "\t\t[features_paths, sequences_paths],\n",
    "\t\tindex=['features_path', 'sequences_path']\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>metadata</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>761380</td>\n",
       "      <td>762036</td>\n",
       "      <td>-</td>\n",
       "      <td>16S ribosomal RNA methyltransferase A</td>\n",
       "      <td>657</td>\n",
       "      <td>&gt;GCF_000213215.1;NC_015518.1;loc:761380,762035...</td>\n",
       "      <td>TTATACATAATTTAATCCCATTAAATTCAATAGTTCTATAACTTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3035</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>1296061</td>\n",
       "      <td>1297560</td>\n",
       "      <td>+</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>1500</td>\n",
       "      <td>&gt;GCF_000213215.1;NC_015518.1;loc:1296061,12975...</td>\n",
       "      <td>TTCCGGTTGATCCTGCCGGACCCGACCGCTATGGGGGTAGGGCTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             assembly genomic_accession    start      end strand  \\\n",
       "1739  GCF_000213215.1       NC_015518.1   761380   762036      -   \n",
       "3035  GCF_000213215.1       NC_015518.1  1296061  1297560      +   \n",
       "\n",
       "                                       name  length  \\\n",
       "1739  16S ribosomal RNA methyltransferase A     657   \n",
       "3035                      16S ribosomal RNA    1500   \n",
       "\n",
       "                                               metadata  \\\n",
       "1739  >GCF_000213215.1;NC_015518.1;loc:761380,762035...   \n",
       "3035  >GCF_000213215.1;NC_015518.1;loc:1296061,12975...   \n",
       "\n",
       "                                               sequence  \n",
       "1739  TTATACATAATTTAATCCCATTAAATTCAATAGTTCTATAACTTGA...  \n",
       "3035  TTCCGGTTGATCCTGCCGGACCCGACCGCTATGGGGGTAGGGCTAA...  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def extract_sequence_from_fasta_record(fasta_record, row):\n",
    "\tsequence = fasta_record.seq[row['start']:row['end']]\n",
    "\tif row['strand'] == '-':\n",
    "\t\tsequence = sequence.reverse_complement()\n",
    "\treturn str(sequence)\n",
    "\n",
    "def add_sequences(df, sequences_filename):\n",
    "\twith gzip.open(sequences_filename, 'rt') as sequences_fasta:\n",
    "\t\tfor fasta_record in SeqIO.parse(sequences_fasta, 'fasta'):\n",
    "\t\t\tfasta_rows = df['genomic_accession'] == fasta_record.id\n",
    "\t\t\tdf.loc[fasta_rows, 'sequence'] = df.loc[fasta_rows].apply(\n",
    "\t\t\t\tlambda row: extract_sequence_from_fasta_record(fasta_record, row),\n",
    "\t\t\t\taxis=1,\n",
    "\t\t\t)\n",
    "\treturn df\n",
    "\n",
    "add_sequences(features_16s_df, sequences_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = features_16s_df.apply(lambda row: row['metadata']+'\\n'+row['sequence']+'\\n', axis=1).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['>GCF_000213215.1;NC_015518.1;loc:761380,762035;strand:-;length657\\nTTATACATAATTTAATCCCATTAAATTCAATAGTTCTATAACTTGACATGGTTTAAACTCCCTCACGCGTTTTTCGCTATTAGATATAAAGCCGCAATATTCTCCTACGTTTTTTAATTTCTTATTCCTAAATCTGCTAACACATTTTAAATACTCGTCTACTTTTGCATTATATGTCCTTTTCCTCATAAAAGTTACTATAATTGAATAAACCTTAGGGCTTGGTGTAAAAGCTGAAGGAGGAATAACTTCGTGAGTTTTAATATCAAAAATATAGTTAAGTAAGAAAGAAATATAAGTAGCATAATTTAATATTTTATCTATAAAATCTTTTTGTAAGATTAAAACTAAACTGATTACTTGATCCAGTTTACTTACTTCCAAGAAAAAATCTTCAGTAATAGAATAAGGTAAGGAGGAGACTATCTGACCTCTCTTTATAGGTAAAAATCTCGCATCTGCGATAATAAGGTTATAAGATTTAAGATATTTAATGAATTTATCATCAATCTCTATGCATAAATCAGGATTAATAACCTTAGAAATATTACCTTTACCGCATCCAACTTCAACGACTGGCTTTATATCATTTTTTACATAGGATGAAAATTTGAAAATAAAAAACTTATCAATCAAAAAATTTTGTGAAAGTTTCA\\n',\n",
       " '>GCF_000213215.1;NC_015518.1;loc:1296061,1297559;strand:+;length1500\\nTTCCGGTTGATCCTGCCGGACCCGACCGCTATGGGGGTAGGGCTAAGCCATGGGAGTCGTACGCCCTCGGGTAAGAGGGCGTGGCGGACGGCTGAGTAACACGTGGCTAACTTACCCTCGGGACCCGGATAACTCCGGGAAACTGGAGCTAATCCGGGGCAGGCGAAGGGTACTGGAACGTCCCTTCGCCTAAAGGGGCATGGGCTATTTCCTGCTCATGCTCGCCCGAGGATAGGGCTGCGGCCCATCAGGCTGTTGGCGGGGTAACGGCCCGCCAAACCGATAACGGGTAGGGGCCGTGAGAGCGGGAGCCCCCAGTTGGGCACTGAGACAAGGGCCCAGGCCCTACGGGGCGCACCAGGCGCGAAACGTCCCCAATGCGGGAAACCGTGAGGGCGCTACCCCCAGTGCTCCCGAAAGGGAGCTTTTCCCCGCTTTAGAACGGCGGGGGAATAAGCGGGGGGCAAGACTGGTGTCAGCCGCCGCGGTAATACCAGCCCCGCGAGTGGTCGGGACGCTTACTGGGCTTAAAGCGCCCGTAGCCGGCCCTGCAAGTCACTGCTTAAAGACCCGGGCTCAACCCGGGAAAGGGCAGTGATACTGCAGGGCTAGGGGGCGGGAGAGGTCGGAGGTACTCCCGGAGTAGGGGCGAAATCCACAGATCCCGGGAGGACCACCAGTGGCGAAAGCGTCCGGCCAGAACGCGCCCGACGGTGAGGGGCGAAAGCCGGGGTAGCAAAAGGGATTAGATACCCCTGTAGTCCCGGCTGTAAACGATGCAGGCTAGGTGTCGCGTAGGCTTTGCGCCTACGCGGTGCCGCAGGGAAACCGGTAAGCCTGCCGCCTGGGGAGTACGCCCGCAAGGGTGAAACTTAAAGGAATTGGCGGGGGAGCACCACAAGGGGTGGAACCTGCGGCTCAATTGGAGTCAACGCCTGGAATCTCACCGGGGGAGACCGCAGGATGACGGCCAGGCTAACGACCTTGCCAGACTCGCGGAGAGGAGGTGCATGGCCGTCGCCAGCTCGTGTTGTGAAATGTCCGGTTAAGTCCGGCAACGAGCGAGACCCCCACCCTTAGTTGGTATCCCGGTCTCCGGATCGGGACCACACTAAGGGGACTGCCGGCGTTAAGCCGGAGGAAGGAGGGGGCCACGGCAGGTCAGCATGCCCCGAAACCCCCGGGCCGCACGCGGGTTACAATGGCAGGGACAGCGGGATTCCGACCCCGAAAGGGGGAGGTAATCCCTCAAACCCTGCCTCAGTTGGGATCGAGGGCTGAAACTCGCCCTCGTGAACGAGGAATCCCTAGTAACCGCACGTCAACAACGTGCGGTGAATACGTCCCTGCTCCTTGCACACACCGCCCGTCGCTCCACCCGAGTGAGGAAGAAGTGAGGCTCCTTGCCCTTCGGGGTGGGGAGTCGAACTTCTTCCTCGCGAGGGGGGAGAAGTCGTAACAAGGTAGCCGTAGGGGAACCTGCGGCTGGATCACCTCAC\\n']"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCF_000213215.1;NC_015518.1;loc:1739     761380\\n3035    1296061\\nName: start, dtype: int64,1739     762035\\n3035    1297559\\nName: end, dtype: int64;strand:+;length1739     657\\n3035    1500\\nName: length, dtype: int64'"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_16s_df['metadata'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from genome_assemblies.summary import (\n",
    "    load_assembly_summary_table,\n",
    "    cleanup_infraspecific_names,\n",
    "    strain_name_from_organism_name,\n",
    "    add_detailed_strain_designation\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>strain_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_900128725.1</td>\n",
       "      <td>na</td>\n",
       "      <td>9</td>\n",
       "      <td>Buchnera aphidicola</td>\n",
       "      <td>BCifornacula</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>2016/11/25</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>Buchnera aphidicola</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_948107705.1</td>\n",
       "      <td>na</td>\n",
       "      <td>9</td>\n",
       "      <td>Buchnera aphidicola</td>\n",
       "      <td>na</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>2023/05/30</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>Buchnera aphidicola</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_008244535.1</td>\n",
       "      <td>na</td>\n",
       "      <td>14</td>\n",
       "      <td>Dictyoglomus thermophilum</td>\n",
       "      <td>PYS_80_B</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2019/09/02</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Dictyoglomus thermophilum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_026418735.1</td>\n",
       "      <td>na</td>\n",
       "      <td>14</td>\n",
       "      <td>Dictyoglomus thermophilum</td>\n",
       "      <td>na</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2022/11/27</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Dictyoglomus thermophilum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_003854875.1</td>\n",
       "      <td>na</td>\n",
       "      <td>17</td>\n",
       "      <td>Methylophilus methylotrophus</td>\n",
       "      <td>D22</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>2018/12/01</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Methylophilus methylotrophus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299256</th>\n",
       "      <td>GCF_030410515.1</td>\n",
       "      <td>na</td>\n",
       "      <td>3058043</td>\n",
       "      <td>Arthrobacter sp. YD4</td>\n",
       "      <td>YD4</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>2023/07/05</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Arthrobacter sp. YD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299257</th>\n",
       "      <td>GCF_030410475.1</td>\n",
       "      <td>na</td>\n",
       "      <td>3058046</td>\n",
       "      <td>Arthrobacter sp. YD2</td>\n",
       "      <td>YD2</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>2023/07/05</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Arthrobacter sp. YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299258</th>\n",
       "      <td>GCF_030410375.1</td>\n",
       "      <td>na</td>\n",
       "      <td>3058168</td>\n",
       "      <td>Zwartia sp. IMCC34845</td>\n",
       "      <td>IMCC34845</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2023/07/05</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Zwartia sp. IMCC34845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299259</th>\n",
       "      <td>GCF_029958925.1</td>\n",
       "      <td>na</td>\n",
       "      <td>3058420</td>\n",
       "      <td>Tropicibacter sp. YMD87</td>\n",
       "      <td>YMD87</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>2023/05/09</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Tropicibacter sp. YMD87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299260</th>\n",
       "      <td>GCF_030412265.1</td>\n",
       "      <td>na</td>\n",
       "      <td>3058424</td>\n",
       "      <td>Fictibacillus sp. CENA-BCM004</td>\n",
       "      <td>CENA-BCM004</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>2023/07/06</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>Fictibacillus sp. CENA-BCM004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>299261 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #assembly_accession refseq_category    taxid  \\\n",
       "0          GCF_900128725.1              na        9   \n",
       "1          GCF_948107705.1              na        9   \n",
       "2          GCF_008244535.1              na       14   \n",
       "3          GCF_026418735.1              na       14   \n",
       "4          GCF_003854875.1              na       17   \n",
       "...                    ...             ...      ...   \n",
       "299256     GCF_030410515.1              na  3058043   \n",
       "299257     GCF_030410475.1              na  3058046   \n",
       "299258     GCF_030410375.1              na  3058168   \n",
       "299259     GCF_029958925.1              na  3058420   \n",
       "299260     GCF_030412265.1              na  3058424   \n",
       "\n",
       "                        organism_name infraspecific_name   assembly_level  \\\n",
       "0                 Buchnera aphidicola       BCifornacula  Complete Genome   \n",
       "1                 Buchnera aphidicola                 na         Scaffold   \n",
       "2           Dictyoglomus thermophilum           PYS_80_B           Contig   \n",
       "3           Dictyoglomus thermophilum                 na           Contig   \n",
       "4        Methylophilus methylotrophus                D22       Chromosome   \n",
       "...                               ...                ...              ...   \n",
       "299256           Arthrobacter sp. YD4                YD4         Scaffold   \n",
       "299257           Arthrobacter sp. YD2                YD2         Scaffold   \n",
       "299258          Zwartia sp. IMCC34845          IMCC34845           Contig   \n",
       "299259        Tropicibacter sp. YMD87              YMD87  Complete Genome   \n",
       "299260  Fictibacillus sp. CENA-BCM004        CENA-BCM004         Scaffold   \n",
       "\n",
       "       seq_rel_date                                           ftp_path  \\\n",
       "0        2016/11/25  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "1        2023/05/30  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "2        2019/09/02  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3        2022/11/27  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "4        2018/12/01  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "...             ...                                                ...   \n",
       "299256   2023/07/05  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "299257   2023/07/05  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "299258   2023/07/05  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "299259   2023/05/09  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "299260   2023/07/06  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "\n",
       "                          strain_name  \n",
       "0                 Buchnera aphidicola  \n",
       "1                 Buchnera aphidicola  \n",
       "2           Dictyoglomus thermophilum  \n",
       "3           Dictyoglomus thermophilum  \n",
       "4        Methylophilus methylotrophus  \n",
       "...                               ...  \n",
       "299256           Arthrobacter sp. YD4  \n",
       "299257           Arthrobacter sp. YD2  \n",
       "299258          Zwartia sp. IMCC34845  \n",
       "299259        Tropicibacter sp. YMD87  \n",
       "299260  Fictibacillus sp. CENA-BCM004  \n",
       "\n",
       "[299261 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_summary_filename = '../data/bacteria_assembly_summary_2023-07-13.txt'\n",
    "assembly_levels = ['Complete Genome', 'Chromosome', 'Scaffold', 'Contig']\n",
    "refseq_levels = ['reference genome', 'representative genome', 'na']\n",
    "\n",
    "as_df = load_assembly_summary_table(assembly_summary_filename)\t\n",
    "as_df['infraspecific_name'] = as_df['infraspecific_name'].apply(cleanup_infraspecific_names)\n",
    "as_df['strain_name'] = as_df['organism_name'].apply(strain_name_from_organism_name)\n",
    "as_df['assembly_level'] = pd.Categorical(\n",
    "\tas_df['assembly_level'],\n",
    "\tcategories=assembly_levels,\n",
    "\tordered=True,\n",
    ")\n",
    "as_df['refseq_category'] = pd.Categorical(\n",
    "\tas_df['refseq_category'],\n",
    "\tcategories=refseq_levels,\n",
    "\tordered=True,\n",
    ")\n",
    "\n",
    "as_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "as_df['strain_name'] = as_df.apply(add_detailed_strain_designation, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>ftp_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_002973605.1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1960156</td>\n",
       "      <td>Abditibacterium utsteinense</td>\n",
       "      <td>LMG 29911</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2018/03/06</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000160075.2</td>\n",
       "      <td>na</td>\n",
       "      <td>592010</td>\n",
       "      <td>Abiotrophia defectiva ATCC 49176</td>\n",
       "      <td>ATCC 49176</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>2013/10/31</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_026783725.1</td>\n",
       "      <td>na</td>\n",
       "      <td>46125</td>\n",
       "      <td>Abiotrophia defectiva</td>\n",
       "      <td>D14035481</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2022/12/09</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_013267415.1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>46125</td>\n",
       "      <td>Abiotrophia defectiva</td>\n",
       "      <td>FDAARGOS_785</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>2020/06/04</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_001815865.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1581061</td>\n",
       "      <td>Abiotrophia sp. HMSC24B09</td>\n",
       "      <td>HMSC24B09</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>2016/10/21</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276905</th>\n",
       "      <td>GCF_934667745.1</td>\n",
       "      <td>na</td>\n",
       "      <td>354118</td>\n",
       "      <td>uncultured Victivallis sp.</td>\n",
       "      <td>na</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2022/04/16</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276906</th>\n",
       "      <td>GCF_946221545.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1161389</td>\n",
       "      <td>uncultured Weeksella sp.</td>\n",
       "      <td>na</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2022/09/12</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276907</th>\n",
       "      <td>GCF_934230365.1</td>\n",
       "      <td>na</td>\n",
       "      <td>253243</td>\n",
       "      <td>uncultured Weissella sp.</td>\n",
       "      <td>na</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2022/04/16</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276908</th>\n",
       "      <td>GCF_947502365.1</td>\n",
       "      <td>na</td>\n",
       "      <td>255433</td>\n",
       "      <td>uncultured Zobellia sp.</td>\n",
       "      <td>na</td>\n",
       "      <td>Contig</td>\n",
       "      <td>2023/01/30</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276909</th>\n",
       "      <td>GCF_918797525.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2355</td>\n",
       "      <td>unidentified bacterial endosymbiont</td>\n",
       "      <td>na</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>2022/06/24</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>276910 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #assembly_accession        refseq_category    taxid  \\\n",
       "0          GCF_002973605.1  representative genome  1960156   \n",
       "1          GCF_000160075.2                     na   592010   \n",
       "2          GCF_026783725.1                     na    46125   \n",
       "3          GCF_013267415.1  representative genome    46125   \n",
       "4          GCF_001815865.1                     na  1581061   \n",
       "...                    ...                    ...      ...   \n",
       "276905     GCF_934667745.1                     na   354118   \n",
       "276906     GCF_946221545.1                     na  1161389   \n",
       "276907     GCF_934230365.1                     na   253243   \n",
       "276908     GCF_947502365.1                     na   255433   \n",
       "276909     GCF_918797525.1                     na     2355   \n",
       "\n",
       "                              organism_name infraspecific_name  \\\n",
       "0               Abditibacterium utsteinense          LMG 29911   \n",
       "1          Abiotrophia defectiva ATCC 49176         ATCC 49176   \n",
       "2                     Abiotrophia defectiva          D14035481   \n",
       "3                     Abiotrophia defectiva       FDAARGOS_785   \n",
       "4                 Abiotrophia sp. HMSC24B09          HMSC24B09   \n",
       "...                                     ...                ...   \n",
       "276905           uncultured Victivallis sp.                 na   \n",
       "276906             uncultured Weeksella sp.                 na   \n",
       "276907             uncultured Weissella sp.                 na   \n",
       "276908              uncultured Zobellia sp.                 na   \n",
       "276909  unidentified bacterial endosymbiont                 na   \n",
       "\n",
       "         assembly_level seq_rel_date  \\\n",
       "0                Contig   2018/03/06   \n",
       "1              Scaffold   2013/10/31   \n",
       "2                Contig   2022/12/09   \n",
       "3       Complete Genome   2020/06/04   \n",
       "4              Scaffold   2016/10/21   \n",
       "...                 ...          ...   \n",
       "276905           Contig   2022/04/16   \n",
       "276906           Contig   2022/09/12   \n",
       "276907           Contig   2022/04/16   \n",
       "276908           Contig   2023/01/30   \n",
       "276909  Complete Genome   2022/06/24   \n",
       "\n",
       "                                                 ftp_path  \n",
       "0       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "2       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "3       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "4       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "...                                                   ...  \n",
       "276905  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "276906  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "276907  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "276908  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "276909  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "\n",
       "[276910 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "as_df = as_df.sort_values(\n",
    "\tby=['assembly_level', 'refseq_category', 'seq_rel_date'],\n",
    "\tascending=[False, False, False]\n",
    ")\n",
    "as_filter_df = (\n",
    "\tas_df\n",
    "\t.groupby('strain_name')\n",
    "\t.first()\n",
    "\t.reset_index(drop=True)\n",
    ")\n",
    "\n",
    "as_filter_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rexmapdb-QKLFbFw_-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
