{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip, os, sys\n",
    "import argparse\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from _include import log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ass_f_dir='features/'\n",
    "ass_s_dir='sequences/'\n",
    "delim=';'\n",
    "match_str='16S ribosomal RNA.*'\n",
    "len_min=500\n",
    "len_max=2000 \n",
    "overwrite=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_processed_assembly_ids(\n",
    "\t\tassembly_df,\n",
    "\t\tfasta_filename,\n",
    "\t\tfasta_delim=';',\n",
    "\t\toverwrite=False\n",
    "\t):\n",
    "\tprocessed_assembly_ids = set([])\n",
    "\tlast_assid = ''\n",
    "\tstart_index = 0\n",
    "\tif os.path.isfile(fasta_filename) and not overwrite:\n",
    "\t\tlog('* output '+fasta_filename+' already exists: ', end='')\n",
    "\t\tlog('appending to file...', end='')\n",
    "\t\twith open(fasta_filename, 'r') as fasta_out:\n",
    "\t\t\tn_processed_ids = 0\n",
    "\t\t\tfor record in SeqIO.parse(fasta_out, 'fasta'):\n",
    "\t\t\t\tlast_assid = record.id.strip().split(fasta_delim)[0]\n",
    "\t\t\t\tprocessed_assembly_ids |= set([last_assid])\n",
    "\t\t\t\tn_processed_ids += 1\n",
    "\t\ttry:\n",
    "\t\t\tlast_checked_index = assembly_df['#assembly_accession'] == last_assid\n",
    "\t\t\tstart_index = assembly_df.index[last_checked_index].tolist()[0]                    \n",
    "\t\texcept:\n",
    "\t\t\tstart_index = 0\n",
    "\t\tlog('OK. Added % 5d sequences.' % (n_processed_ids))\n",
    "\treturn (processed_assembly_ids, start_index)\n",
    "\n",
    "def load_gff_feature_table(filename):\n",
    "\tcolumns = {\n",
    "\t\t'assembly': str,\n",
    "\t\t'genomic_accession':str,\n",
    "\t\t'start': int,\n",
    "\t\t'end': int, \n",
    "\t\t'strand': str,\n",
    "\t\t'name': str,\n",
    "\t\t'attributes': str\n",
    "\t}\n",
    "\treturn (\n",
    "\t\tpd.read_csv(\n",
    "\t\t\tfilename,\n",
    "\t\t\tsep='\\t',\n",
    "\t\t\tcompression='gzip',\n",
    "\t\t\tusecols=list(columns.keys()),\n",
    "\t\t\tdtype=columns,\n",
    "\t\t\tkeep_default_na=False,\n",
    "\t\t)\t\t\t\t\t\n",
    "\t)\n",
    "\n",
    "def build_file_path(ftp_path, base_path, type):\n",
    "\tsuffix = {'features': 'feature_table.txt', 'sequences': 'genomic.fna'}\n",
    "\treturn os.path.join(\n",
    "\t\tbase_path,\n",
    " \t\tos.path.basename(ftp_path) + '_' + suffix[type] + '.gz',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ass_sub = '../data/archaea_assembly_summary_filter_2023-07-13.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>ftp_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009428885.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_009729015.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_003201835.2</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_009729545.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1455</th>\n",
       "      <td>GCF_001316265.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1456</th>\n",
       "      <td>GCF_001316285.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1457</th>\n",
       "      <td>GCF_001748385.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1458</th>\n",
       "      <td>GCF_902786095.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>GCF_948444175.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1460 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #assembly_accession                                           ftp_path\n",
       "0        GCF_009428885.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1        GCF_009729015.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "2        GCF_003201835.2  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "3        GCF_000213215.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "4        GCF_009729545.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "...                  ...                                                ...\n",
       "1455     GCF_001316265.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1456     GCF_001316285.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1457     GCF_001748385.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...\n",
       "1458     GCF_902786095.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...\n",
       "1459     GCF_948444175.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...\n",
       "\n",
       "[1460 rows x 2 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fa_out_name = '../test.fa'\n",
    "\n",
    "assembly_df = pd.read_csv(\n",
    "\tass_sub,\n",
    "\tsep='\\t',\n",
    "\tusecols=['#assembly_accession', 'ftp_path'],\n",
    "\tdtype={'#assembly_accession': str, 'ftp_path': str}\n",
    ")\n",
    "\n",
    "processed_assembly_ids, start_index = add_processed_assembly_ids(assembly_df, fa_out_name)\n",
    "assembly_df = assembly_df[~assembly_df['#assembly_accession'].isin(processed_assembly_ids)]\n",
    "assembly_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_paths(ftp_path, base_paths):\n",
    "\tpath_prefix = os.path.basename(ftp_path)\n",
    "\tfeatures_paths = os.path.join(base_paths['features'], path_prefix + '_feature_table.txt.gz')\n",
    "\tsequences_paths = os.path.join(base_paths['sequences'], path_prefix + '_genomic.fna.gz')\n",
    "\treturn pd.Series(\n",
    "\t\t[features_paths, sequences_paths],\n",
    "\t\tindex=['features_path', 'sequences_path']\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>features_path</th>\n",
       "      <th>sequences_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009428885.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009428885.1_ASM942888v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009428885.1_ASM942888v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_009729015.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009729015.1_ASM972901v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009729015.1_ASM972901v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_003201835.2</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_003201835.2_ASM320183v2_f...</td>\n",
       "      <td>../data/sequences/GCF_003201835.2_ASM320183v2_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000213215.1_ASM21321v1_fe...</td>\n",
       "      <td>../data/sequences/GCF_000213215.1_ASM21321v1_g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_009729545.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009729545.1_ASM972954v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009729545.1_ASM972954v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1455</th>\n",
       "      <td>GCF_001316265.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_001316265.1_ASM131626v1_f...</td>\n",
       "      <td>../data/sequences/GCF_001316265.1_ASM131626v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1456</th>\n",
       "      <td>GCF_001316285.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_001316285.1_ASM131628v1_f...</td>\n",
       "      <td>../data/sequences/GCF_001316285.1_ASM131628v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1457</th>\n",
       "      <td>GCF_001748385.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_001748385.1_ASM174838v1_f...</td>\n",
       "      <td>../data/sequences/GCF_001748385.1_ASM174838v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1458</th>\n",
       "      <td>GCF_902786095.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>../data/features/GCF_902786095.1_Rumen_uncultu...</td>\n",
       "      <td>../data/sequences/GCF_902786095.1_Rumen_uncult...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>GCF_948444175.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>../data/features/GCF_948444175.1_SRR17622767_b...</td>\n",
       "      <td>../data/sequences/GCF_948444175.1_SRR17622767_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1460 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #assembly_accession                                           ftp_path  \\\n",
       "0        GCF_009428885.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1        GCF_009729015.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "2        GCF_003201835.2  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3        GCF_000213215.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "4        GCF_009729545.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "...                  ...                                                ...   \n",
       "1455     GCF_001316265.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1456     GCF_001316285.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1457     GCF_001748385.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1458     GCF_902786095.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "1459     GCF_948444175.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "\n",
       "                                          features_path  \\\n",
       "0     ../data/features/GCF_009428885.1_ASM942888v1_f...   \n",
       "1     ../data/features/GCF_009729015.1_ASM972901v1_f...   \n",
       "2     ../data/features/GCF_003201835.2_ASM320183v2_f...   \n",
       "3     ../data/features/GCF_000213215.1_ASM21321v1_fe...   \n",
       "4     ../data/features/GCF_009729545.1_ASM972954v1_f...   \n",
       "...                                                 ...   \n",
       "1455  ../data/features/GCF_001316265.1_ASM131626v1_f...   \n",
       "1456  ../data/features/GCF_001316285.1_ASM131628v1_f...   \n",
       "1457  ../data/features/GCF_001748385.1_ASM174838v1_f...   \n",
       "1458  ../data/features/GCF_902786095.1_Rumen_uncultu...   \n",
       "1459  ../data/features/GCF_948444175.1_SRR17622767_b...   \n",
       "\n",
       "                                         sequences_path  \n",
       "0     ../data/sequences/GCF_009428885.1_ASM942888v1_...  \n",
       "1     ../data/sequences/GCF_009729015.1_ASM972901v1_...  \n",
       "2     ../data/sequences/GCF_003201835.2_ASM320183v2_...  \n",
       "3     ../data/sequences/GCF_000213215.1_ASM21321v1_g...  \n",
       "4     ../data/sequences/GCF_009729545.1_ASM972954v1_...  \n",
       "...                                                 ...  \n",
       "1455  ../data/sequences/GCF_001316265.1_ASM131626v1_...  \n",
       "1456  ../data/sequences/GCF_001316285.1_ASM131628v1_...  \n",
       "1457  ../data/sequences/GCF_001748385.1_ASM174838v1_...  \n",
       "1458  ../data/sequences/GCF_902786095.1_Rumen_uncult...  \n",
       "1459  ../data/sequences/GCF_948444175.1_SRR17622767_...  \n",
       "\n",
       "[1460 rows x 4 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_df[['features_path', 'sequences_path']] = assembly_df['ftp_path'].apply(\n",
    "    build_paths,\n",
    "    base_paths = {\n",
    "        'features': '../data/'+ass_f_dir,\n",
    "        'sequences': '../data/'+ass_s_dir,\n",
    "    }\n",
    ")\n",
    "assembly_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def file_exists(filename):\n",
    "\treturn os.path.exists(filename) and os.path.getsize(filename) > 0\n",
    "\n",
    "def filter_missing_files(assembly_df):\n",
    "\tassembly_df['features_exist'] = assembly_df['features_path'].apply(file_exists)\n",
    "\tassembly_df['sequences_exist'] = assembly_df['sequences_path'].apply(file_exists)\n",
    "\treturn (\n",
    "\t\tassembly_df\n",
    "\t\t.query('features_exist == True & sequences_exist == True')\n",
    "\t\t.drop(columns=['features_exist', 'sequences_exist'])\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>features_path</th>\n",
       "      <th>sequences_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009428885.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009428885.1_ASM942888v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009428885.1_ASM942888v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_009729015.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009729015.1_ASM972901v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009729015.1_ASM972901v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_003201835.2</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_003201835.2_ASM320183v2_f...</td>\n",
       "      <td>../data/sequences/GCF_003201835.2_ASM320183v2_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000213215.1_ASM21321v1_fe...</td>\n",
       "      <td>../data/sequences/GCF_000213215.1_ASM21321v1_g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_009729545.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_009729545.1_ASM972954v1_f...</td>\n",
       "      <td>../data/sequences/GCF_009729545.1_ASM972954v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1445</th>\n",
       "      <td>GCF_014962245.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_014962245.1_ASM1496224v1_...</td>\n",
       "      <td>../data/sequences/GCF_014962245.1_ASM1496224v1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446</th>\n",
       "      <td>GCF_000092185.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000092185.1_ASM9218v1_fea...</td>\n",
       "      <td>../data/sequences/GCF_000092185.1_ASM9218v1_ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1447</th>\n",
       "      <td>GCF_000148385.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000148385.1_ASM14838v1_fe...</td>\n",
       "      <td>../data/sequences/GCF_000148385.1_ASM14838v1_g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1449</th>\n",
       "      <td>GCF_001316005.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_001316005.1_ASM131600v1_f...</td>\n",
       "      <td>../data/sequences/GCF_001316005.1_ASM131600v1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1450</th>\n",
       "      <td>GCF_000190315.1</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>../data/features/GCF_000190315.1_ASM19031v1_fe...</td>\n",
       "      <td>../data/sequences/GCF_000190315.1_ASM19031v1_g...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1450 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #assembly_accession                                           ftp_path  \\\n",
       "0        GCF_009428885.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1        GCF_009729015.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "2        GCF_003201835.2  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3        GCF_000213215.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "4        GCF_009729545.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "...                  ...                                                ...   \n",
       "1445     GCF_014962245.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1446     GCF_000092185.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1447     GCF_000148385.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1449     GCF_001316005.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1450     GCF_000190315.1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "\n",
       "                                          features_path  \\\n",
       "0     ../data/features/GCF_009428885.1_ASM942888v1_f...   \n",
       "1     ../data/features/GCF_009729015.1_ASM972901v1_f...   \n",
       "2     ../data/features/GCF_003201835.2_ASM320183v2_f...   \n",
       "3     ../data/features/GCF_000213215.1_ASM21321v1_fe...   \n",
       "4     ../data/features/GCF_009729545.1_ASM972954v1_f...   \n",
       "...                                                 ...   \n",
       "1445  ../data/features/GCF_014962245.1_ASM1496224v1_...   \n",
       "1446  ../data/features/GCF_000092185.1_ASM9218v1_fea...   \n",
       "1447  ../data/features/GCF_000148385.1_ASM14838v1_fe...   \n",
       "1449  ../data/features/GCF_001316005.1_ASM131600v1_f...   \n",
       "1450  ../data/features/GCF_000190315.1_ASM19031v1_fe...   \n",
       "\n",
       "                                         sequences_path  \n",
       "0     ../data/sequences/GCF_009428885.1_ASM942888v1_...  \n",
       "1     ../data/sequences/GCF_009729015.1_ASM972901v1_...  \n",
       "2     ../data/sequences/GCF_003201835.2_ASM320183v2_...  \n",
       "3     ../data/sequences/GCF_000213215.1_ASM21321v1_g...  \n",
       "4     ../data/sequences/GCF_009729545.1_ASM972954v1_...  \n",
       "...                                                 ...  \n",
       "1445  ../data/sequences/GCF_014962245.1_ASM1496224v1...  \n",
       "1446  ../data/sequences/GCF_000092185.1_ASM9218v1_ge...  \n",
       "1447  ../data/sequences/GCF_000148385.1_ASM14838v1_g...  \n",
       "1449  ../data/sequences/GCF_001316005.1_ASM131600v1_...  \n",
       "1450  ../data/sequences/GCF_000190315.1_ASM19031v1_g...  \n",
       "\n",
       "[1450 rows x 4 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_missing_files(assembly_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>attributes</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3035</th>\n",
       "      <td>GCF_000213215.1</td>\n",
       "      <td>NC_015518.1</td>\n",
       "      <td>1296061</td>\n",
       "      <td>1297560</td>\n",
       "      <td>+</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td></td>\n",
       "      <td>1500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             assembly genomic_accession    start      end strand  \\\n",
       "3035  GCF_000213215.1       NC_015518.1  1296061  1297560      +   \n",
       "\n",
       "                   name attributes  length  \n",
       "3035  16S ribosomal RNA               1500  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find all genomic_accession for all 16S rRNA\n",
    "\n",
    "# Load GFF, find 16S records, filter by length\n",
    "features_filename = assembly_df.loc[3, 'features_path']\n",
    "\n",
    "def load_16s_records(features_filename, length_thresholds={'min': 500, 'max': 2000}):\n",
    "\tmatch_str = '16S ribosomal RNA(?!.*methyl).*'\n",
    "\tdf = load_gff_feature_table(features_filename)\n",
    "\tfeatures_df = df[df['name'].str.contains(match_str)].copy()\n",
    "\tfeatures_df['length'] = features_df['end'] - features_df['start'] + 1\n",
    "\tfeatures_df = features_df.query(\n",
    "\t\tf'length >= {length_thresholds[\"min\"]} & length <= {length_thresholds[\"max\"]}'\n",
    "\t)\n",
    "\treturn features_df\n",
    "\n",
    "features_df = load_16s_records(features_filename=)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/sequences/GCF_000213215.1_ASM21321v1_genomic.fna.gz'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences_filename = assembly_df.loc[3, 'sequences_path']\n",
    "sequences_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with gzip.open(sequences_filename, 'rt') as sequences_fasta:\n",
    "\tfor fasta_record in SeqIO.parse(sequences_fasta, 'fasta'):\n",
    "\t\tfeatures_16s_df = features_df[features_df['genomic_accession'] == fasta_record.id]\n",
    "\t\tif not features_16s_df.empty:\n",
    "\t\t\tfor j in range(len(feature_record_df)):\n",
    "\t\t\t\tif ass_f_sub_sub.iloc[j]['strand'] == '-':\n",
    "\t\t\t\t\tseq = str(record.seq[start:end].reverse_complement())\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tseq = str(record.seq[start:end])\n",
    "\t\t\t\tfa_out.write('>' + ass_id + delim + record.id + delim + 'loc:' + \\\n",
    "\t\t\t\t\t\t\t\tstr(start) + ',' + str(end-1) + delim + 'strand:' + \\\n",
    "\t\t\t\t\t\t\t\tass_f_sub_sub.iloc[j]['strand'] + delim + 'length:' + \\\n",
    "\t\t\t\t\t\t\t\tstr(length) + '\\n' + \\\n",
    "\t\t\t\t\t\t\t\tseq + '\\n')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rexmapdb-QKLFbFw_-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
